File: 프로그래머스/2/12978. 배달/배달.cpp
#include <bits/stdc++.h>
using namespace std;

#define INF 500'001

vector<pair<int,int>> edges[51]; // 인덱스: src, first: dest, second: dist
int distList[51]; // 0번으로 부터의 거리
priority_queue<pair<int,int>> pq; // 거리, dest
bool chk[51];
int ans;

int solution(int n, vector<vector<int>> road, int k) {
    
    // init
    for(int i=0;i<road.size();i++){
        vector<int> iter = road[i];
        edges[iter[0]-1].emplace_back(iter[1]-1,iter[2]);
        edges[iter[1]-1].emplace_back(iter[0]-1,iter[2]);
    }
    
    for(int i=0;i<n;i++){
        distList[i] = INF;
    }
    pq.push({0,0});
    
    // solve
    while(!pq.empty()){
        pair<int,int> cur = pq.top();
        pq.pop();
        int dist = cur.first;
        int dest = cur.second;

        distList[dest] = dist;
        
        for(pair<int,int> next : edges[dest]){
            if(dist + next.second > distList[next.first]) continue;
            pq.push(make_pair(dist + next.second, next.first));
        }
    }
    
    // for(int i=0;i<n;i++){
    //     cout << distList[i] << ' ';
    // }cout << endl;
    for(int i=0;i<n;i++){
        if(distList[i]<=k)ans++;
    }
    
    return ans;
}

File: 프로그래머스/2/12978. 배달/배달.java
import java.util.*;

class Solution {
    int n;
    int INF = Integer.MAX_VALUE;
    public class Edge implements Comparable<Edge>{
        int dest; // 목적지
        int cost; // 비용
        public Edge(int d,int c){
            this.dest = d;
            this.cost = c;
        }
        public int compareTo(Edge e){
            return Integer.compare(this.cost, e.cost);
        }
    }
    List<Edge>[] graph;
    int[] dist;
    public void init(int N, int[][] road){
        graph = new List[N];
        dist = new int[N];
        
        for(int i=0;i<N;i++){
            graph[i] = new ArrayList();
        }
        for(int i=0;i<road.length;i++){
            int src = road[i][0]-1;
            int dest = road[i][1]-1;
            int cost = road[i][2];
            graph[src].add(new Edge(dest,cost));
            graph[dest].add(new Edge(src,cost));
        }
        for(int i=0;i<N;i++){
            dist[i]=INF;
        }
        
    }
    
    public void djikstra(){
        PriorityQueue<Edge> pq = new PriorityQueue();
        pq.offer(new Edge(0,0));
        dist[0] = 0;
        while(!pq.isEmpty()){
            Edge cur = pq.poll();
            for(Edge e : graph[cur.dest]){
                if(dist[e.dest]<=cur.cost + e.cost)continue;
                dist[e.dest]=cur.cost + e.cost;
                pq.offer(new Edge(e.dest,dist[e.dest]));
            }
        }
        
    }
    public int solution(int N, int[][] road, int K) {
        init(N,road);
        int ans = 0;
        djikstra();
        for(int i=0;i<N;i++){
            if(dist[i]>K)continue;
            ans++;
        }
        return ans;
    }
}